flang/lib/Optimizer/Transforms/FirLoopResultOpt.cpp

```diff
@@ -0,0 +1,108 @@
+//===- FirLoopResultOpt.cpp - Optimization pass for fir loops    ------ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "flang/Optimizer/Dialect/FIRDialect.h"
+#include "flang/Optimizer/Dialect/FIROps.h"
+#include "flang/Optimizer/Dialect/FIRType.h"
+#include "flang/Optimizer/Transforms/Passes.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+
+#define DEBUG_TYPE "flang-fir-result-opt"
+
+namespace {
+
+class LoopResultRemoval : public mlir::OpRewritePattern<fir::DoLoopOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LoopResultRemoval(mlir::MLIRContext *c)
+      : OpRewritePattern(c) {}
+  mlir::LogicalResult
+  matchAndRewrite(fir::DoLoopOp loop,
+                  mlir::PatternRewriter &rewriter) const override {
+    for (auto r : loop.getResults()) {
+      if (valueUseful(r))
+        return mlir::failure();
+    }
+    auto &loopOps = loop.getBody()->getOperations();
+    auto newLoop = rewriter.create<fir::DoLoopOp>(
+        loop.getLoc(), loop.lowerBound(), loop.upperBound(), loop.step());
+    rewriter.startRootUpdate(newLoop.getOperation());
+    rewriter.startRootUpdate(loop.getOperation());
+    newLoop.getBody()->getOperations().splice(
+        --newLoop.getBody()->end(), loopOps, loopOps.begin(), --loopOps.end());
+    loop.getInductionVar().replaceAllUsesWith(newLoop.getInductionVar());
+    rewriter.finalizeRootUpdate(loop.getOperation());
+    rewriter.finalizeRootUpdate(newLoop.getOperation());
+    for (auto r : loop.getResults()) {
+      eraseAllUses(r, rewriter);
+    }
+    rewriter.eraseBlock(loop.getBody());
```

Comment:
Does this do a dropAllReferences() on the Region for you?

Reply:
yes, `dropAllReferences` looks better. Thank you!